rllib/utils/replay_buffers/storage.py

```diff
@@ -0,0 +1,503 @@
+        self._eviction_started = state["_eviction_started"]
+        self._num_timesteps_added = state["_num_timesteps_added"]
+        self._num_timesteps = state["_num_timesteps"]
+        self._size_bytes = state["_size_bytes"]
+        self._hit_count = np.zeros(self.capacity, dtype=np.int64)
+
+    @ExperimentalAPI
+    def __len__(self) -> int:
+        return self._num_items
+
+    @ExperimentalAPI
+    def __iter__(self) -> Iterator[SampleBatchType]:
+        for i in range(len(self)):
+            yield self[i]
+
+    @ExperimentalAPI
+    def __getitem__(self, i: int) -> SampleBatchType:
+        if i >= len(self) or i < 0:
+            raise IndexError('Buffer index out of range')
+        idx = self._get_internal_index(i)
+        self._hit_count[idx] += 1
+        return self._get(idx)
+
+    @ExperimentalAPI
+    def __setitem__(self, i: int, item: SampleBatchType) -> None:
+        if i >= len(self) or i < 0:
+            raise IndexError('Buffer index out of range')
+        if not self.eviction_started:
+            raise RuntimeError(
+                "Assigning items to an index is only allowed "
+                "after eviction has been started. Use .add(item) instead."
+            )
+        idx = self._get_internal_index(i)
+        drop_item = self._del(idx)
+        if drop_item.count < item.count:
+            logger.warning(
+                "New item consists of more timesteps than "
+                "the replaced item. This violates storage capacity."
+            )
+        self._evicted_hit_stats.push(self._hit_count[idx])
+        self._num_timesteps -= drop_item.count
+        self._size_bytes -= drop_item.size_bytes()
+        self._hit_count[idx] = 0
+        self._num_timesteps_added += item.count
+        self._num_timesteps += item.count
+        self._size_bytes += item.size_bytes()
+        self._set(idx, item)
+
+    @ExperimentalAPI
+    def add(self, item: SampleBatchType) -> None:
+        """ Add a new item to the storage. The index of the new item
+        will be automatically assigned. Moreover, old items may be
+        automatically dropped with respect to the storage's capactiy.
+
+        Args:
+            item: Item (batch) to add to the storage.
+        """
+        if item.count > self.capacity:
+            logger.warning(
+                "The batch to be added consists of {} timesteps "
+                "which is larger than the storage capacity of {}. "
+                "Therefore, the batch has not been added.".format(
+                    item.count, self.capacity
+                )
+            )
+            return
+        
+        self._num_timesteps_added += item.count
+        self._num_timesteps += item.count
+        self._size_bytes += item.size_bytes()
+
+        # Drop old items.
+        # May require multiple drops if newly added item
+        # contains more timesteps than the old items.
+        while self._num_timesteps > self.capacity:
```

Comment:
@sven1977 I'm super fine with starting to treat capacity as /timesteps. What do you say?

Reply:
How would you handle having to store whole batches, and having to store sequences for rnn's?

Reply:
I just adopted the definition of capacity from the current ReplayBuffer implementation, i.e. it has already been timesteps before (see [here](https://github.com/ray-project/ray/blob/2bce07d4ce935a54e142e05c46d7f4595d1995ad/rllib/utils/replay_buffers/replay_buffer.py#L179)). The storage does only monitor the number of timesteps, not the order or structure of experiences. That is left up to the ReplayBuffer. Overall, Storage does not change the behavior of ReplayBuffer regarding the storage units.

Reply:
Awesome! Really cool PR @fwitter!